qualtran/_infra/data_types.py

```diff
@@ -810,6 +810,18 @@ def assert_valid_classical_val_array(
         if np.any(val_array >= 2**self.bitsize):
             raise ValueError(f"Too-large classical values encountered in {debug_str}")
 
+    def montgomery_inverse(self, xm, n, p):
+        return ((pow(xm, p - 2, p)) * pow(2, 2 * n, p)) % p
+
+    def montgomery_product(self, xm, ym, n, p):
+        return (xm * ym * pow(2, n * (p - 2), p)) % p
+
+    def montgomery_to_uint(self, xm, n, p):
+        return (xm * pow(2, n * (p - 2), p)) % p
+
+    def uint_to_montgomery(self, x, n, p):
+        return (x * pow(2, n, p)) % p
```

Comment:
sorry, you've opened a can-of-worms with these methods :)

Right now, these can be turned into `@staticmethod`s since they don't use `self`. Why don't they use `self`? Shouldn't `n` be `self.bitsize`? Why is the prime `p` not an attribute on the data type class? 

 - please open and link-with-a-todo and issue that says the datatype should learn these parameters
 - in the interim, change these to static methods and document(!) the input parameters

Reply:
Opened https://github.com/quantumlib/Qualtran/issues/1471 for adding modulus p to QMontgomeryUInt(). I think this would require too many changes for the scope of this PR. Happy to do the follow up PR after my project is completed.

Documented + tested - didn't make static cause I opted to use self.bitsize.